### DISCUSSION TITLE
license checkout failed

### DISCUSSION THREAD

**Discussion by Spclaim** [ansys/pyansys-geometry#754] [Q&A] [answered]
Hi,
I setup the docker image according to the documentation. But during startup of the container I receive the message License checkout failed.

When I try to run a python demo script from the documentation I receive the following error.

ansys.geometry.core.errors.GeometryExitedError: Geometry service connection terminated: Exception was thrown by handler.

Unclear if the license checkout is root cause of the thrown python error

help would be appreciated.

**Comment by RobPasMue** [comment]
Hi @Spclaim - could you please share with us the ``docker run`` command and the docker container logs that you get?

You can hide out the license server IP/DNS for security purposes, for sure =)

**Reply by Spclaim** [reply]
Thank you for your response. 

I forgot to mention it is docker for windows. Installed ansys version 2023R2.

` docker logs --details pedantic_blackburn

 Licensing: Setting ANSYS241_DIR=C:\app\SpaceClaim.DirectModelingService.Internal.dll\ansys
 Licensing: Using license server 1055@xxxxxxxxx
 Licensing: License checkout failed
 ApiServerAddin: ApiServerAddin: License checkout failed.

 License checkout failed.

 DMS started.
 gRPC services subscription done.
 ApiServerAddin: ApiServerAddin: Descriptors added. Descriptors count: 19.
 ApiServerAddin: ApiServerAddin: Services added. Services count: 18.
 ApiServerAddin: ApiServerAddin: Starting server routine: new port added to the server: Grpc.Core.ServerPort.
 Server started. Listening on endpoint: 0.0.0.0:50051`


`docker run -p 700:50051 --env-file .env ghcr.io/ansys/geometry:windows-latest`

env-file:

`LICENSE_SERVER=xxxxxx
SERVER_ENDPOINT=127.0.0.1:50051
ENABLE_TRACE=1
LOG_LEVEL=2`

**Reply by RobPasMue** [reply]
No problem! Seems like the license checkout is failing, meaning that it is not connecting to the license server. We have seen some issues with the license servers when:

* Working inside a VPN
* Deploying the license server on localhost (i.e. on your own machine)

If the license server server is in a VPN, and you are connected to the VPN, it is more of a Docker network issue. Your Docker engine is not going through your VPN properly and does not resolve your host. I recommend you restart your Docker engine in that case and try again.

Another common "fix" is to force the dns resolution when spawning the container. Try passing in the flag ``--dns "8.8.8.8"`` to force it to go through Google's DNS.

Let me know if any of these solutions work!

**Reply by Spclaim** [reply]
I thought that too. But when running 

Docker exec pedantic_blackburn ping licenseserverhostname

Pinging works .

Also when attached to the container:

C:\app\licensingclient\winx64>ansysli_util -statli 2325@xxxxxxxxx
Messages file C:\Program Files\ANSYS Inc\Shared Files\Licensing\language\en-us\ansysli_msgs.xml does not exist.



Attempting to connect to ansysli_server running on 2325@xxxxxxx
Connection Established.  Querying for additional information...
        ANSYSLI_PORT = 2325

C:\app\licensingclient\winx64>ansysli_util -statli 2325@xxxxxxx
Messages file C:\Program Files\ANSYS Inc\Shared Files\Licensing\language\en-us\ansysli_msgs.xml does not exist.



Attempting to connect to ansysli_server running on 2325@xxxxxxxxxxx...
Connection Established.  Querying for additional information...
        ANSYSLI_PORT = 2325
            Hostname = xxxxxxxxxxx
                  IP = 192.168.230.207
          Process Id = 3000
        Full Version = 2023.2.0 (20230523) for winx64.  Updated at Ansys Release 2023 R2.
             Startup = NORMAL STARTUP 2023/09/16 02:08:05
      Active Clients = 4
         Max Clients = 10
       Total Clients = 9654
   WB Usage Tracking = ON
  DP Reserve Support = ON
     Service Status  = No problems found
Service Status Message : The Ansys LM service and the Licensing Interconnect are using the same exe.
     Monitor Status    : Monitor Running
Monitor Version        : Ansys Licensing Interconnect version 2023.1.0 (20230523) for winx64.  Updated at Ansys Release 2023 R2.
** ERROR: Unable to retrieve feature availability for 2325@xxxxxxx. **

**Reply by RobPasMue** [reply]
Looks like your License Server is exposing port 2325. We are assuming port 1055 by default as you can see in the first log

Try passing in your LICENSE_SERVER variable as ``2325@XXXXXXXXX`` and see if it works.

**Reply by Spclaim** [reply]
Unfortunately this also does not work. I checked standard ansys licensing client tool and the port is 1055. 

I will try to install on a different machine and see if it is this machine...

**Reply by Spclaim** [reply]
Is the license type we have irrelevant ? we have ansys mechanical enterprise.

on a different machine it also does not work :(

**Reply by RobPasMue** [reply]
You should have a Discovery/SpaceClaim license - otherwise it will not work. Sorry, should have clarified this before

**Reply by Spclaim** [reply]
We can use Spaceclaim / Discovery because it is included in the mechanical enterprise license. I was under the assumption that this is sufficient and we do not need an extra license only valid for spaceclaim / discovery. Sorry for the confusion. This is a bit counter-intuitive.

**Reply by RobPasMue** [reply]
Let me check with the team and get back to you. I am not sure of the details inside the licensing check.

**Reply by RobPasMue** [reply]
Before anything else...

* Did you try passing in the ``--dns="8.8.8.8"`` flag? Did this work?

I have received confirmation that the Ansys Mechanical Enterprise license is not sufficient to use the Ansys Geometry Service. However, you are saying that you have access to both SpaceClaim and Discovery so, let's do another trial:

* Do you have Ansys 23R2 SP1? If that's the case, continue to the next step. Otherwise, you will need to install it.
* Let's try running PyAnsys Geometry against SpaceClaim and Discovery. Bear in mind that interaction with Discovery is limited (you will only be able to launch Discovery with ``ansys-geometry-core==0.3.2``. Starting on 24R1 you will be able to perform all operations. In 23R2 SP1 you can perform most operations with SpaceClaim though:

  * Launch SC from PyAnsys Geometry
  
    ```py
    from ansys.geometry.core import launch_modeler_with_spaceclaim
    
    modeler = launch_modeler_with_spaceclaim()
    print(modeler)
    ```

  * Launch Discovery from PyAnsys Geometry
  
    ```py
    from ansys.geometry.core import launch_modeler_with_discovery
    
    modeler = launch_modeler_with_discovery()
    print(modeler)
    ```

Your Ansys products should start fine if your license is valid. Let's try verifying it.

**Reply by RobPasMue** [reply]
You may find more information on how to launch the local installation of Ansys SpaceClaim and Discovery here: https://geometry.docs.pyansys.com/version/0.3/getting_started/local/index.html

**Comment by Spclaim** [comment]
Hi thank you,

no dns 8.8.8.8 is not reachable from company network. We have our own DNS server, but this also didn’t work.

I have 23R2 not sure about SP1.

But I can start both SC and Discovery from python. Only issue is that I must close it manually.

**Reply by RobPasMue** [reply]
Okay, could you try running some of the online examples for PyAnsys Geometry and see if they work?

Also, if you run

```py
modeler.close()
```

Do you still have to close it manually?

**Reply by RobPasMue** [reply]
You will have to run an actual modelling example. For example: https://geometry.docs.pyansys.com/version/0.3/examples/02_sketching/dynamic_sketch_plane.html

**Reply by Spclaim** [reply]
Really appreciate your help. 

modeler.close() works properly . 

the sketch tutorial worked with a pyvista plot of the created sketch. nice. 

the  parametric gear sketch does not work and the single body with material assignment also does not work.

**Reply by RobPasMue** [reply]
Hi @Spclaim - yes, support for SpaceClaim and Discovery from PyAnsys Geometry, at least on 23R2 is limited. Certain operations might fail. The intended way to operate with PyAnsys Geometry is against the Ansys Geometry Service. Support will be improved starting on 24R1 and we will try to make most of our APIs compatible with the Ansys standalone products.

**Reply by RobPasMue** [reply]
Just to be clear - all PyAnsys Geometry APIs will work against the Ansys Geometry Service binaries you found on the release. But we do not ensure that all APIs work against the Ansys standalone products on 23R2.

**Comment by RobPasMue** [comment, marked as answer]
After further investigation we have reached the main issue on the licensing - the Ansys Geometry Service (headless and dockerized solution) only works with a Discovery Modeling license. If you have an Ansys Mechanical Enterprise license, for example, you will be able to use PyAnsys Geometry with SpaceClaim or Discovery, but not the Ansys Geometry Service.

On 23R2 SP1 it will only work with SpaceClaim due to some issues with Discovery. Also, support to Ansys standalone products (SpaceClaim, Discovery) is limited from the PyAnsys Geometry API. Certain operations might fail since the intended interaction of this library is with the Ansys Geometry Service directly.

Starting on 24R1, we will provide larger support for the Ansys standalone products (Discovery, SpaceClaim). For PyAnsys Geometry version compatibility with Ansys products please refer to https://geometry.docs.pyansys.com/version/stable/getting_started/compatibility.html

Thanks @Spclaim for opening this discussion. I will mark this response as an answer for future reference.

**Reply by RobPasMue** [reply]
We will transfer this discussion to the FAQ section in our documentation.

**Comment by torayeff** [comment]
@RobPasMue regarding the license. I am on Ansys Startup Program. I assume I should have the Discovery Modeling license. How can I verify if I have it and how do I enable it?

**Reply by RobPasMue** [reply]
Hi @torayeff - please reach out to your Ansys contact and request this information if possible. On my side I have no information on the licenses available for the Startup program sadly, although I would expect it to be available
